test-data/unit/check-optional.test

```diff
@@ -832,3 +832,133 @@ main:4: error: Argument 1 to "asdf" has incompatible type "List[str]"; expected
+
+def f1(b: bool) -> Optional[int]:
+    if b:
+        z = 10
+        reveal_type(z)  # N: Revealed type is "builtins.int"
+    else:
+        z = None
+        reveal_type(z)  # N: Revealed type is "None"
+    reveal_type(z)  # N: Revealed type is "Union[builtins.int, None]"
+    return z
+
+def f2(b: bool) -> int:
+    if b:
+        z = 10
+    else:
+        z = None
+    return z  # E: Incompatible return value type (got "Optional[int]", expected "int")
+
+def f3(b: bool) -> int:
+    # XXX: This one is a little questionable! Maybe we *do* want to allow this?
+    z = 10
+    if b:
+        z = None  # E: Incompatible types in assignment (expression has type "None", variable has type "int")
+    return z
+
+def f4() -> Optional[int]:
+    z = 10
+    z = None  # E: Incompatible types in assignment (expression has type "None", variable has type "int")
+
+    return z
+
+def f5() -> None:
+    z = 10
+
+    def f() -> None:
+        nonlocal z
+        z = None  # E: Incompatible types in assignment (expression has type "None", variable has type "int")
+
+def f6(b: bool) -> None:
+    if b:
+        z = 10
+    else:
+        z = 11
+
+    def f() -> None:
+        nonlocal z
+        z = None  # E: Incompatible types in assignment (expression has type "None", variable has type "int")
+
+def f7(b: bool) -> None:
+    if b:
+        z = 10
+    else:
+        z = 11
+
+    z = None  # E: Incompatible types in assignment (expression has type "None", variable has type "int")
+
+def f8(b: bool, c: bool) -> Optional[int]:
+    if b:
+        if c:
+            z = 10
+        else:
+            z = 11
+    else:
+        z = None
+    return z
+
+def f9(b: bool) -> None:
+    if b:
+        z: int = 10
+    else:
+        z = None  # E: Incompatible types in assignment (expression has type "None", variable has type "int")
+
+def f10(b: bool) -> None:
+    z: int
+    if b:
+        z = 10
+    else:
+        z = None  # E: Incompatible types in assignment (expression has type "None", variable has type "int")
+
+def f11(b: bool, c: bool) -> None:
+    if b:
+        z = 10
+    elif c:
+        z = 30
+    else:
+        z = None
+
+def f12(b: bool, a: Any) -> None:
+    if b:
+        z = a
+    else:
+        z = None
+    reveal_type(z)  # N: Revealed type is "Any"
+
+def f13(b: bool, a: Any) -> None:
+    if b:
+        try:
+            z = f2(True)
+        except Exception:
+            raise RuntimeError
+    else:
+        z = None
+
+def f14(b: bool, a: Any) -> None:
+    if b:
+        with a:
+            z = 10
+    else:
+        z = None
+
+def f15() -> None:
+    try:
+        z = f2(True)
+    except Exception:
+        z = None
+    reveal_type(z)  # N: Revealed type is "Union[builtins.int, None]"
+
+def f16(z: Any) -> None:
+    for x in z:
+        if x == 0:
+            y = 50
+            break
+    else:
+        y = None
+    reveal_type(y)  # N: Revealed type is "Union[builtins.int, None]"
+
```

Comment:
Ideas for additional test cases below. I'm not sure what the semantics should be in some of these cases, but at least it makes sense to have *some* well-defined semantics for these. Omitted most `reveal_type(z)` etc. for brevity.

Case 1:

```
if b:
    z = 2
elif c:
    z = None
elif d:
    z = 3
```

Case 2:

```
if c:
    x = []
else:
    x = None
```

Case 3:

```
if b:
    z = 4
else:
    if c:
        z = 5
    else:
        z = None
```

Case 4:

Use this at module top level, or within class body.

Case 5:

```
if b:
    z = 5
else:
    z = None
    if c:
        z = 6
```

Case 6:

```
if b:
    x = []
    x.append(1)
else:
    x = None
```

Case 7:

```
if b:
    x: Any = 5
else:
    x = None
reveal_type(x)
```

Case 8:

[Tricky, optional] Use at module top level, and have a forward reference that causes two passes over the top level during type checking.

Case 9:

[Tricky, optional] Have a forward reference where the type of the reference target is not ready yet within a function between the first and second assignment. This causes the function to be deferred.

Case 10:

Define an attribute using `self.x = 5` / `self.x = None`. I think that it's not safe to allow this, and we may need to reject this, since the attribute type is visible outside the method (see below for an example where this might cause trouble).

Case 11:

[Very tricky, optional] Define an attribute, and cause the method to be deferred after the initial `self.x` assignment. Now another method refers to the `x` attribute, and it should not see the non-optional type.